lib/kast.js

```diff
@@ -67,6 +67,56 @@ const joinMatches = (match1, match2) => {
 const antimatch = (pattern, term, path = []) => {
   if (pattern.node == KVARIABLE) {
     return [];
+  } else if (pattern.node == KAPPLY && term.node == KAPPLY && pattern.label == term.label && pattern.label == "_Map_") {
+    // gather keys lhs
+    const lhs = pattern.args
+      .filter(pair => pair.node !== "KVariable")
+      .map(pair => [format(pair.args[0]), pair.args[1]])
+    const rhs = term.args
+      .filter(pair => pair.node !== "KVariable")
+      .map(pair => [format(pair.args[0]), pair.args[1]])
+    const lhs_keys = lhs.map(pair => pair[0]);
+    const rhs_keys = rhs.map(pair => pair[0]);
+    // TODO - introduce a type arg which states if arg is missing or rejected
+    const missing_keys = rhs
+      .filter(([key, v]) => !lhs_keys.includes(key))
+    const rejected_keys = lhs
+      .filter(([key, v]) => !rhs_keys.includes(key))
+    const matching_keys = lhs
+      .filter(([key, v]) => rhs_keys.includes(key))
+
+    const matching = matching_keys
+      .map(([key, pattern_]) => antimatch(pattern_, rhs.find(el => el[0] == key)[1], path.concat( key.replace(/\n\s*/g, "") )))
+      .reduce((a,b) => a.concat(b), []);
+    const missing = missing_keys
+      .map(([key, pattern_]) => [path.concat( key.replace(/\n\s*/g, "") ), format(pattern_), ""] )
+    const rejected = rejected_keys
+      .map(([key, pattern_]) => [path.concat( key.replace(/\n\s*/g, "") ), "", format(pattern_)] )
+    return matching.concat(missing).concat(rejected);
+  } else if (pattern.node == KAPPLY && term.node == KAPPLY && pattern.label == term.label && pattern.label == "_Set_") {
+    const reduceSet = set => {
+      if(set.label == "_Set_") {
+        return reduceSet(set.args[1]).concat(set.args[0].args[0])
+      } else {
+        return [set];
+      }
+    };
+    const lhs_set = reduceSet(pattern)
+      .map(set_item => format(set_item))
+    const rhs_set = reduceSet(pattern)
+      .map(set_item => format(set_item))
+
+    const missing_items = rhs_set
+      .filter(set_item => !lhs_set.includes(set_item))
+    const rejected_items = lhs_set
+      .filter(set_item => !rhs_set.includes(set_item))
+
+    if(missing_items.length == 0 && rejected_items.length == 0) {
+      return [];
+    } else {
+      return [path.concat("Set"), lhs_set, rhs_set];
+    }
+
```

Comment:
what is going on here? what is an antimatch?

Reply:
**what is antimatch**
antimatch is a special case of antiunification where you only worry to match/unify the left side, instead of both.
generally in unification you have two terms with variables (e.g. `a(12, X) and a(Y, 4)`) and you looking for a substitution for all the variables so that they are equal (e.g. `X=4 and Y=12`)
in anti unification you have two terms that differ (e.g. `a(12, 4) and a(12, foo)`) and try to get a geneneralized terms (e.g. `a(12, X)`) which has variables where both differ, together with substitution for each side: (e.g. `X=4` for the lhs and `X=foo` for rhs)
**why its relevant**
there is a target antiunification view when you hit `T` (uppercase) which tells you at every term where it is different from the target term you want to reach.
Its handy to debug failing proofs

antiunification: https://en.wikipedia.org/wiki/Anti-unification_(computer_science)

Reply:
it was implemented before but in a dirty way, not respecting maps and sets
fixed here the maps and sets part

Reply:
Very interesting. Thanks Denis! :heart: